ROX-23789: Central sends scan configd on sensor connection

jschnath opened the pull request:
## Description

Central side counterpart to https://github.com/stackrox/stackrox/pull/11051

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

### Here I tell how I validated my change

~~No testing beyond~~ confirming that the messages sent are of the type expected in Sensor as seen in https://github.com/stackrox/stackrox/pull/11051.

Added unit tests
E2e tests work in progress

lvalerom commented on central/sensor/service/connection/connection_impl.go:
```diff
@@ -500,6 +502,56 @@ func (c *sensorConnection) getClusterConfigMsg(ctx context.Context) (*central.Ms
+			profiles = append(profiles, profile.ProfileName)
+		}
+		scanConfigRequest := central.ApplyComplianceScanConfigRequest{
+			ScanRequest: &central.ApplyComplianceScanConfigRequest_ScheduledScan_{
```
This is the incorrect message. You can check the expected message in sensor's side or the integration tests

jschnath replied:
Done, thanks

lvalerom commented on central/sensor/service/connection/connection_impl.go:
```diff
@@ -500,6 +502,56 @@ func (c *sensorConnection) getClusterConfigMsg(ctx context.Context) (*central.Ms
+				ScheduledScan: &central.ApplyComplianceScanConfigRequest_ScheduledScan{
+					ScanSettings: &central.ApplyComplianceScanConfigRequest_BaseScanSettings{
+						ScanName:       scanConfig.ScanConfigName,
+						StrictNodeScan: true,
```
Why did you set this to true?

jschnath replied:
I saw it set like that in another place where I copied the scansetting from. I'll remove it.

lvalerom replied:
The field is ignored in sensor but I think, regardless of whether is ignored or not, you should use the value from the scan config.

lvalerom reviewed (changes requested):
The overall idea is fine but this won't work (see the comments I added to pinpoint where the issues are).

Also, the unit tests are failing after your changes in the `service_impl.go`, you need to update them.

lvalerom reviewed (changes requested):
I think it's important to add e2e tests and do some manual testing to make sure the feature is working as expected. 

Also the unit tests still need to be updated after your changes.

jschnath commented on central/sensor/service/connection/connection_impl.go:
```diff
@@ -500,6 +503,60 @@ func (c *sensorConnection) getClusterConfigMsg(ctx context.Context) (*central.Ms
+	var reformatedConfigs []*central.ApplyComplianceScanConfigRequest
+	for _, scanConfig := range scanConfigs {
+		var profiles []string
+		for _, profile := range scanConfig.Profiles {
```
I did not know we had those, I'll use them now. I wasn't used to having a public attribute and also a Get(), for me it's usually been an accessible attribute or Get() *and* Set().

jschnath commented on central/sensor/service/connection/connection_impl.go:
```diff
@@ -500,6 +503,60 @@ func (c *sensorConnection) getClusterConfigMsg(ctx context.Context) (*central.Ms
+	for _, scanConfig := range scanConfigs {
+		var profiles []string
+		for _, profile := range scanConfig.Profiles {
+			profiles = append(profiles, profile.ProfileName)
```
Done

jschnath commented on central/sensor/service/connection/connection_impl.go:
```diff
@@ -500,6 +503,60 @@ func (c *sensorConnection) getClusterConfigMsg(ctx context.Context) (*central.Ms
+			profiles = append(profiles, profile.ProfileName)
+		}
+		var cron string
+		cron, err = schedule.ConvertToCronTab(scanConfig.Schedule)
```
Done

jschnath commented on central/sensor/service/connection/connection_impl.go:
```diff
@@ -500,6 +503,60 @@ func (c *sensorConnection) getClusterConfigMsg(ctx context.Context) (*central.Ms
+			ScanRequest: &central.ApplyComplianceScanConfigRequest_UpdateScan{
+				UpdateScan: &central.ApplyComplianceScanConfigRequest_UpdateScheduledScan{
+					ScanSettings: &central.ApplyComplianceScanConfigRequest_BaseScanSettings{
+						ScanName: scanConfig.ScanConfigName,
```
Done

jschnath commented:
Added a test

jschnath commented:
/retest

jschnath commented on central/complianceoperator/v2/scanconfigurations/service/service_impl.go:
```diff
@@ -47,7 +47,7 @@ var (
 	})
 
-	configNameRegexp = regexp.MustCompile(`^[a-z0-9][a-z0-9.-]*[a-z0-9]?$`)
+	configNameRegexp = regexp.MustCompile(`^[a-z0-9][a-z0-9.-]*$`)
```
@lvalerom 
I've shortened this Regex because the shorter version is logically equivalent to the longer one, but this makes me wonder if the Regex maybe was incorrect in the first place. Do you know the rules for config names we are trying to enforce here?

jschnath commented:
/help

jschnath commented:
/test compliance-e2e

jschnath commented:
/test ocp-4-15-compliance-e2e-tests

jschnath commented:
/retest

jschnath commented:
/retest

jschnath commented:
/retest

lvalerom commented:
@jschnath The test e2e test you added is not running. Ping me again when they are please.

jschnath commented:
Huh, did I run the wrong target? I thought they were included in the ocp compliance-e2e-tests which passed. I'll run all tests named compliance then and see if that includes it.
/test ocp-4-12-compliance-e2e-tests
/test ocp-4-15-compliance-e2e-tests
/test ocp-4-candidate-compliance-e2e-tests
/test ocp-4-preview-compliance-e2e-tests
/test ocp-stable-scanner-v4-compliance-e2e-tests

jschnath commented:
One of the compliance e2e tests is failing, ci/prow/ocp-stable-scanner-v4-compliance-e2e-tests, but it the actual tests failing in there are "OOM Check / Check for scanner-v4-db OOM" and "Pod Restarts / scanner". I guess rerunning them wouldn't hurt, but I assume my test has ran and passed now(?) I don't see any other failures in our compliance tests. @lvalerom 

I'll also rerun what might be flakes.
/retest

lvalerom commented:
> One of the compliance e2e tests is failing, ci/prow/ocp-stable-scanner-v4-compliance-e2e-tests, but it the actual tests failing in there are "OOM Check / Check for scanner-v4-db OOM" and "Pod Restarts / scanner". I guess rerunning them wouldn't hurt, but I assume my test has ran and passed now(?) I don't see any other failures in our compliance tests. @lvalerom

Answered in slack but just in case someone stumbles upon this in the future: The problem is not the CI jobs not running or passing, the problem is that the new test is ignored in the step and it never gets executed. I see you already figured out the problem. Let's see if the test passes now 🙂

jschnath commented:
/retest

jschnath commented:
Can't figure out why cluster creation failed, let's see if it happens again.

/retest

jschnath commented:
Let's see if the gcloud cluster creation issue persists

/retest

jschnath commented:
/retest

jschnath commented:
/retest

jschnath commented:
/retest

jschnath commented:
@lvalerom ocp-4-12-compliance-e2e-tests are still failing with the gcloud credentials problem Tom is working on, but if you look at the ocp-4-15-compliance-e2e-tests you'll see that my e2e test is part of the target and passing.

lvalerom commented on central/sensor/service/connection/connection_test.go:
```diff
@@ -57,6 +87,63 @@ func (c *mockServer) Recv() (*central.MsgFromSensor, error) {
+
+	caps := set.NewSet[centralsensor.SensorCapability]([]centralsensor.SensorCapability{centralsensor.SendDeduperStateOnReconnect}...)
+
+	mgrMock.EXPECT().GetCluster(ctx, gomock.Any()).Return(&storage.Cluster{}, true, nil).AnyTimes()
```
I think this should be called only twice.

jschnath replied:
I can do that, but then this test will have to be touched again if the number of GetCluster calls changes, I didn't think it mattered for this test how often that was called.

lvalerom commented on central/sensor/service/connection/connection_test.go:
```diff
@@ -57,6 +87,63 @@ func (c *mockServer) Recv() (*central.MsgFromSensor, error) {
+	caps := set.NewSet[centralsensor.SensorCapability]([]centralsensor.SensorCapability{centralsensor.SendDeduperStateOnReconnect}...)
+
+	mgrMock.EXPECT().GetCluster(ctx, gomock.Any()).Return(&storage.Cluster{}, true, nil).AnyTimes()
+	deduper.EXPECT().GetSuccessfulHashes().Return(map[string]uint64{"deployment:1": 0}).Times(1)
```
You don't need this.

jschnath replied:
Yeah, part of the deduper code.

lvalerom commented on central/sensor/service/connection/connection_test.go:
```diff
@@ -57,6 +87,63 @@ func (c *mockServer) Recv() (*central.MsgFromSensor, error) {
+		}
+	}
+
+	s.Equal(1, len(complianceRequests))
+	ScanConfigs := complianceRequests[0].GetSyncScanConfigs().GetScanConfigs()
+	s.Equal(1, len(ScanConfigs))
+	UpdateScan := ScanConfigs[0].GetUpdateScan()
+	s.Equal("TestConfigName", UpdateScan.GetScanSettings().GetScanName())
+	Profiles := UpdateScan.GetScanSettings().GetProfiles()
+	s.Equal(1, len(Profiles))
+	s.Equal("TestProfileName", Profiles[0])
+	s.Equal("2 1 * * *", UpdateScan.GetCron())
```
Use `Require` before accessing the slice in a fixed position. Otherwise some of these might panic.

Also, I think it would be better to iterate through the scan configs and assert on all of them. I know that you have it fixed to just one but this test would fail if in the future someone changes the `scanConfig` variable.

jschnath replied:
Sure, I've refactored it to go over all configs and added a require there 👍

lvalerom commented on central/sensor/service/connection/connection_test.go:
```diff
@@ -170,6 +257,7 @@ func (s *testSuite) TestSendDeduperStateIfSensorReconciliation() {
 				sensorEventHandler: eventHandler,
 				sensorHello:        hello,
 				hashDeduper:        deduper,
+				scanSettingDS:      s.scanConfigDS,
```
This shouldn't be needed in this test.

jschnath replied:
When I remove the DS from the test it panics with `invalid memory address or nil pointer dereference`, I assume because it needs a DS to send the scanconfigs to.

lvalerom commented on central/sensor/service/connection/connection_test.go:
```diff
@@ -265,14 +354,16 @@ func (s *testSuite) TestSendsAuditLogSyncMessageIfEnabledOnRun() {
-		clusterMgr: mgrMock,
+		clusterID:     clusterID,
+		clusterMgr:    mgrMock,
+		scanSettingDS: s.scanConfigDS,
```
This shouldn't be needed in this test.

jschnath replied:
I think it is necessary to have the DS in every test now, otherwise they panic now that central sends a scan config to the datastore

lvalerom commented on central/sensor/service/connection/connection_test.go:
```diff
@@ -366,7 +457,9 @@ func (s *testSuite) TestDelegatedRegistryConfigOnRun() {
 		clusterMgr:                 mgrMock,
 		delegatedRegistryConfigMgr: deleRegMgr,
 		imageIntegrationMgr:        iiMgr,
+		scanSettingDS:              s.scanConfigDS,
```
This shouldn't be needed in this test.

jschnath replied:
The DS is now needed because on connection central sends the scan configuration there

lvalerom reviewed (changes requested):
@jschnath The e2e tests are not working as they should. Sensor is not being scaled down (you can see that in the artifacts). Also, while checking that the scan configs are inserted in the db, you are not asserting that the ScanSettings/ScanSettingBindings are created in the cluster, nor are you testing updates to the scan config.

jschnath commented on central/sensor/service/connection/connection_impl.go:
```diff
@@ -500,6 +505,59 @@ func (c *sensorConnection) getClusterConfigMsg(ctx context.Context) (*central.Ms
+				UpdateScan: &central.ApplyComplianceScanConfigRequest_UpdateScheduledScan{
+					ScanSettings: &central.ApplyComplianceScanConfigRequest_BaseScanSettings{
+						ScanName: scanConfig.GetScanConfigName(),
+						Profiles: profiles,
```
Last we talked about ScanName, Profiles and Cron were the only relevant fields from `storage.ComplianceOperatorScanConfigurationV2` that could be applied to a `central.ApplyComplianceScanConfigRequest`.

I just checked again, and the only other field the two types share is ID, which I thought should not be the same. Should it be? Did I miss something else?

jschnath commented on central/sensor/service/connection/connection_impl.go:
```diff
@@ -697,6 +755,14 @@ func (c *sensorConnection) Run(ctx context.Context, server central.SensorService
 
 	metrics.IncrementSensorConnect(c.clusterID, c.sensorHello.GetSensorState().String())
```
Makes sense, done

jschnath commented on central/sensor/service/connection/connection_impl.go:
```diff
@@ -697,6 +755,14 @@ func (c *sensorConnection) Run(ctx context.Context, server central.SensorService
 
 	metrics.IncrementSensorConnect(c.clusterID, c.sensorHello.GetSensorState().String())
 
+	scanMsg, err := c.getScanConfigurationMsg(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "unable to get scan config for %q", c.clusterID)
+	}
+	if err := server.Send(scanMsg); err != nil {
+		return errors.Wrapf(err, "unable to sync config to cluster %q", c.clusterID)
+	}
```
I'm not sure what you mean, could you elaborate?

jschnath commented on central/sensor/service/connection/connection_test.go:
```diff
@@ -57,6 +87,63 @@ func (c *mockServer) Recv() (*central.MsgFromSensor, error) {
+	eventHandler := newSensorEventHandler(&storage.Cluster{}, "", pipeline, nil, &stopSig, deduper, nil)
+
+	sensorMockConn := &sensorConnection{
+		//clusterID:          fixtureconsts.Cluster1,
```
Removed it now, yeah it was left over from testing what works.

jschnath commented on central/sensor/service/connection/connection_test.go:
```diff
@@ -57,6 +87,63 @@ func (c *mockServer) Recv() (*central.MsgFromSensor, error) {
+		sendC:              make(chan *central.MsgToSensor),
+	}
+
+	caps := set.NewSet[centralsensor.SensorCapability]([]centralsensor.SensorCapability{centralsensor.SendDeduperStateOnReconnect}...)
```
Because the other unit tests I looked at used it too, but you're right, it's not necessary here. I'm removing it.

jschnath commented:
@lvalerom Thanks again for the review.

I've addressed your comments regarding the unit test, good feedback. But as for the e2e test I'm a bit confused, You say that:
> Sensor is not being scaled down (you can see that in the artifacts)
and I've looked through the artifacts but I cannot find that information there. I probably don't know what to look for, how did you see this? Either way I'll still have to work on the e2e tests some more it seems.

jschnath commented:
PR branch was forked and finished by @lvalerom in https://github.com/stackrox/stackrox/pull/11827, closing this one
